web-crawler/src/main/java/com/google/sps/webcrawler/WebCrawler.java

```diff
@@ -0,0 +1,274 @@
+  private RelevancyChecker relevancyChecker;
+  // Mappings of (website robots.txt URL, the next allowed time to access, in milliseconds) for
+  // respecting the required crawl delay.
+  private Map<String, Long> nextAccessTimes = new HashMap<>();
+
+  /** 
+   * Constructs a {@code WebCrawler} instance.
+   *
+   * @throws IOException if {@code RelevancyChecker} instantiation fails, such as because of lack
+   *     of permission to access required libraries.
+   */
+  public WebCrawler() throws IOException {
+    this.relevancyChecker = new RelevancyChecker();
+  }
+
+  /** 
+   * Compiles news articles for the candidate with the specified {@code candidateName} and
+   * {@code candidateId}:
+   * 1. Obtains news article URLs from Google Custom Search.
+   * 2. Checks for permission to web-scrape.
+   * 3. Web-scrapes if permitted.
+   * 4. Extracts content from HTML structure.
+   * 5. Checks content relevancy to the candidate of interest.
+   * 6. Processes content.
+   * 7. Stores processed content in the database.
+   */
+  public void compileNewsArticle(String candidateName, String candidateId) {
+    List<URL> urls = getUrlsFromCustomSearch(candidateName);
+    for (URL url : urls) {
+      Optional<NewsArticle> potentialNewsArticle = scrapeAndExtractHtml(url);
+      if (!potentialNewsArticle.isPresent()) {
+        continue;
+      }
+      NewsArticle newsArticle = potentialNewsArticle.get();
+      if (!relevancyChecker.isRelevant(newsArticle, candidateName)) {
+        continue;
+      }
+      NewsArticle processedNewsArticle = NewsContentProcessor.process(newsArticle);
+      storeInDatabase(candidateId, processedNewsArticle);
+    }
+  }
+
+  // @TODO [Might adopt in {@code compileNewsArticle}.]
+  /** 
+   * Controls the web scraping frequency by delaying the @{code WebCrawler} for the maximum amount
+   * of time as required by webpages encountered so far. This method is for frequency-tuning
+   * purposes only. {@code WebCrawler} will confirm that the required crawl delay is met, before
+   * actually moving forward with the web-scraping.
+   */
+  // private void waitForMaxCrawlDelay() {
+  //   if (nextAccessTimes.isEmpty()) {
+  //     return;
+  //   }
+  //   long timeToDelay = Collections.max(nextAccessTimes.values()) - System.currentTimeMillis();
+  //   try {
+  //     TimeUnit.MILLISECONDS.sleep(timeToDelay);
+  //   } catch (InterruptedException e) {}
+  // }
+
+  // @TODO [Test with Google Custom Search. Extract other metadata.]
+  /** 
+   * Searches for {@code candidateName} in the Google Custom Search engine and finds URLs of news
+   * articles. Returns an empty list if no valid URLs are found.
+   *
+   * @see <a href=
+   *    "https://hc.apache.org/httpcomponents-client-ga/httpclient/examples/org/apache/" +
+   *    "http/examples/client/ClientWithResponseHandler.java">Code reference</a>
+   */
+  public List<URL> getUrlsFromCustomSearch(String candidateName) {
+    // For testing purposes.
```

Comment:
You have "For testing purposes" all over this function. You should make it clear whether this function is meant only to be used in tests OR if it's for hardcoded purposes

Reply:
Apologies for the confusion! This is because I haven't been given access to a custom search engine yet, so the function will throw errors. However, I still wanted the web crawler to be tested from the beginning to the end, so I incorporated this function. In order to do so, I have to add hard-coded URLs, for the next function to have URLs to scrape.

Reply:
I added descriptions for the above.